components/eamxx/src/share/util/eamxx_time_interpolation.cpp

```diff
@@ -0,0 +1,343 @@
+}
+/*-----------------------------------------------------------------------------------------------*/
+/* A function to perform time interpolation using data from all the fields stored in the local
+ * field managers.
+ * Conducts a simple linear interpolation between two points using
+ *        y* = w*y0 + (1-w)*y1
+ * where w = (t1-dt)/(t1-t0)
+ * Input:
+ *   time_in - A timestamp to interpolate onto.
+ * Output
+ *   A map of (field name) and (Field) pairs such that each field is the interpolated values and
+ *   the map makes it possible to quickly query individual outputs.
+ */
+std::map<std::string,Field> TimeInterpolation::perform_time_interpolation(const TimeStamp& time_in)
+{
+  // Declare the output map
+  std::map<std::string,Field> interpolated_fields;
+
+  // If data is handled by files we need to check that the timestamps are still relevant
+  if (m_file_data_triplets.size()>0) {
+    check_and_update_data(time_in);
+  }
+
+  // Gather weights for interpolation.  Note, timestamp differences are integers and we need a
+  // real defined weight.
+  const Real w_num = m_time1 - time_in;
+  const Real w_den = m_time1 - m_time0;
+  const Real weight = w_num/w_den;
+
+  // Cycle through all stored fields and conduct the time interpolation
+  for (auto name : m_field_names)
+  {
+    const auto& field0 = m_fm_time0->get_field(name);
+    const auto& field1 = m_fm_time1->get_field(name);
+    Field field_out    = field0.clone();
+    field_out.update(field1,1.0-weight,weight);
+    interpolated_fields.emplace(name,field_out);
+  }
+
+  // Return map of interpolated fields
+  return interpolated_fields;
+}
+/*-----------------------------------------------------------------------------------------------*/
+/* Function which registers a field in the local field managers.
+ * Input:
+ *   field_in - Is a field with the appropriate dimensions and metadata to match the interpolation
+ * Output:
+ *   None
+ */
+void TimeInterpolation::add_field(const Field& field_in)
+{
+  // First check that we haven't already added a field with the same name.
+  const auto name = field_in.name();
+  EKAT_REQUIRE_MSG(!m_fm_time0->has_field(name) and !m_fm_time1->has_field(name),
+		  "Error!! TimeInterpolation:add_field, field + " << name << " has already been added." << "\n");
+
+  // Clone the field for each field manager to get all the metadata correct.
+  auto field0 = field_in.clone();
+  auto field1 = field_in.clone();
+  m_fm_time0->add_field(field0);
+  m_fm_time1->add_field(field1);
+  m_field_names.push_back(name);
+}
+/*-----------------------------------------------------------------------------------------------*/
+/* Function to shift the data of a single field from time1 to time0
+ * Input:
+ *   name - The name of the field to be shifted.
+ */
+void TimeInterpolation::shift_data(const std::string& name)
+{
+        auto  field0 = m_fm_time0->get_field(name);
+  const auto& field1 = m_fm_time1->get_field(name);
+  field0.deep_copy(field1);
```

Comment:
Rather than a deep copy, how about doing `std::swap(field0,field1)`? This should do a (soft) swap of internal metadata pointers, which means just a handful of pointers...

Reply:
I'm not sure this will work, since the field manager will still be storing the swapped pointers.  Thus when the data is read as input it will end up going to field0 when it is supposed to go to field1.  Note - when I do the swap the unit test fails at the timestep where shift data should have been called.

Reply:
True, the fm returns a copy, not a reference. But if you call `get_field_ptr`, then you can swap the pointers content.